feat: first step refactor resources setting page (#1337)

lstocchi opened the pull request:
### What does this PR do?

This is the first piece of the resource page refactoring.
I'm opening this PR so you can start looking at it. There are still many missing pieces (e.g you cannot create any new machine, some button doesn't perform any action ...) that will be added in the next PR.

It contains 
- different color background + left navbar
- the first piece of the new resources page
  - the podman section that shows all podman machines, their resources usage and the basic button/action start/restart/stop/delete

### Screenshot/screencast of this PR

![podman-resource-1](https://user-images.githubusercontent.com/49404737/221908815-595715fe-4c95-4087-89e0-45e5544ed5c9.gif)

### What issues does this PR fix or reference?

#1337 

### How to test this PR?

1. start it
2. go to the settings/resources page

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -183,12 +183,19 @@ declare module '@tmpwip/extension-api' {
     endpoint: ContainerProviderConnectionEndpoint;
     lifecycle?: ProviderConnectionLifecycle;
     status(): ProviderConnectionStatus;
+    machineInfo?: MachineInfo;
```
we can't add this information
 cpu, memory, diskSize are properties bound to the provider
https://github.com/containers/podman-desktop/blob/main/extensions/podman/package.json#L28-L49
it's per provider

lstocchi replied:
I don't get it. 
Those information are fetched by running this command
` const machineListOutput = await execPromise(getPodmanCli(), ['machine', 'list', '--format', 'json']);`
and that is related to the machines, not the provider. 
if i execute it on my cli i get
```
[
    {
        "Name": "test2",
        "Default": false,
        "Created": "2023-03-01T16:24:02.8246355+01:00",
        "Running": true,
        "Starting": false,
        "LastUp": "2023-03-01T16:26:45.4546807+01:00",
        "Stream": "35",
        "VMType": "wsl",
        "CPUs": 8,
        "Memory": "456134656",
        "DiskSize": "542113792",
        "Port": 56535,
        "RemoteUsername": "user",
        "IdentityPath": "C:\\Users\\Luca\\.ssh\\test2"
    },
    {
        "Name": "podman-machine-default",
        "Default": true,
        "Created": "2023-03-01T16:23:09.9128478+01:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2023-03-01T16:23:09.9128478+01:00",
        "Stream": "35",
        "VMType": "wsl",
        "CPUs": 0,
        "Memory": "0",
        "DiskSize": "507510784",
        "Port": 56524,
        "RemoteUsername": "user",
        "IdentityPath": "C:\\Users\\Luca\\.ssh\\podman-machine-default"
    }
]
```
if i display those cpus/memory/disk values it seems reasonable.
Different things on linux which the `updateMachines` it's never called but i didn't manage it yet.

benoitf replied:
a provider is "generic" so it can provide things or not
the list of properties that is provided is specified by the provider itself
you're listing me `podman machine info` command so it's specific to podman, not to all providers.

> if i display those cpus/memory/disk values it seems reasonable. They are specific infos about 2 podman machines.

it's podman specific, thus it can't be in the API
but it's exposed by properties of the podman extension

as previously linked https://github.com/containers/podman-desktop/blob/main/extensions/podman/package.json#L28-L49
here, podman extension says that it can provide cpu, memory, disk and some other stuff

but for Docker for example we don't provide this information 

so as of today, the resource component is "smart" and only displays stuff that extension is providing

You can display the "cpu, memory and disk" but those are not part of the API, but are properties that are exposed
so if CPU is exposed then you'll be able to display that information, etc.

benoitf commented on packages/main/src/plugin/api/provider-info.ts:
```diff
@@ -36,6 +37,7 @@ export interface ProviderContainerConnectionInfo {
   };
   lifecycleMethods?: LifecycleMethod[];
   type: 'docker' | 'podman';
+  machineInfo?: MachineInfo;
```
```suggestion

```

benoitf commented on packages/extension-api/src/extension-api.d.ts:
```diff
@@ -183,12 +183,19 @@ declare module '@tmpwip/extension-api' {
     socketPath: string;
   }
 
+  export interface MachineInfo {
+    cpus: number;
+    memory: number;
+    diskSize: number;
+  }
+
```
```suggestion


```

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,187 @@
+                        <span class="my-auto text-gray-500 ml-1 font-bold text-[9px]">{container.status.toUpperCase()}</span>
+                        {/if}
+                    </div>
+                    {#if container.machineInfo}
+                    <div class="flex mt-3 {container.status !== 'started' ? 'text-gray-500' : ''}">
+                        <div class="mr-4">
+                            <div class="text-[9px]">vCPUs</div>
+                            <div class="text-xs">{container.machineInfo.cpus}</div>
+                        </div>
+                        <div class="mr-4">
+                            <div class="text-[9px]">MEMORY</div>
+                            <div class="text-xs">{container.machineInfo.memory.toFixed(1)}MB</div>
+                        </div>
+                        <div class="mr-4">
+                            <div class="text-[9px]">DISK LIMIT</div>
+                            <div class="text-xs">{container.machineInfo.diskSize.toFixed(2)}GB</div>
+                        </div>
```
it seems that you're removing all the generic aspect of the providers
we have properties that declare their type, range, etc
here you're hardcoding what is generic

benoitf reviewed (dismissed):
We have providers having a generic way of providing properties for their configuration.
We need to keep that.
We can't have 'machineInfo' which is tied to a specific type which would be 'podman' and for Windows and macOS as for example on Linux there is no machine at all.

Each extension is able to provide custom properties that we can configure (could be cpu, memory, or whatever)
It's providing what can be done at runtime information and 'creation time'/factory (like creating an instance)

https://github.com/containers/podman-desktop/blob/main/extensions/podman/package.json#L33-L83

Each properties have a scope and depending on the screen in Podman Desktop we need to display these properties based on the current scope

lstocchi commented:
@benoitf i think i've fixed it properly. Please give it a look when you can. Thanks!!

benoitf commented on extensions/podman/src/extension.ts:
```diff
@@ -307,6 +327,8 @@ async function registerProviderFor(provider: extensionApi.Provider, machineInfo:
     },
   };
 
+  containerProviderConnections.set(machineInfo.name, containerProviderConnection);
```
hello, it looks like we never remove the field from the Map `containerProviderConnections`

benoitf commented on packages/main/src/plugin/configuration-registry.ts:
```diff
@@ -42,6 +42,12 @@ export interface IConfigurationChangeEvent {
   scope: containerDesktopAPI.ConfigurationScope;
 }
 
+export interface IProviderContainerConfigurationPropertyRecordedSchema extends IConfigurationPropertyRecordedSchema {
+  value?: any;
+  container: string;
+  providerId: string;
+}
+
```
Could you explain why we need this schema ?
The schema declare the definition, not the values

benoitf commented on packages/renderer/src/lib/preferences/PreferencesPage.svelte:
```diff
@@ -37,50 +38,51 @@ onMount(async () => {
-    <Route path="/extensions">
-      <ExtensionList />
-    </Route>
+<div class="flex h-full px-3 py-3 bg-zinc-900">
```
Probably it could be a separate PR removing the header for settings (it's not specific to the resources page)
so it could be easier to check/merge ensuring the new background color is fitting well with all settings

lstocchi replied:
Opened https://github.com/containers/podman-desktop/pull/1670

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,302 @@
+  providerInfos.subscribe(providerInfosValue => {
+    providers = providerInfosValue;
+    providers.forEach(provider => {
+      provider.containerConnections.forEach(container => {
```
what happens if a provider has been removed ?

lstocchi replied:
The provider will not be displayed anymore as the providers array has been updated and the page gets refreshed, no?

benoitf replied:
My question was more: should we display EmptyScreen/stuff to tell the user to do some actions in order to have resources

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,302 @@
+  .filter(property => property.scope === 'ContainerConnection')
+  .sort((a, b) => a.id.localeCompare(b.id));
+
+let tmpProviderContainerConfiguration: IProviderContainerConfigurationPropertyRecordedSchema[] = [];
```
looks like the interface you declared globally `IProviderContainerConfigurationPropertyRecordedSchema` is only used there in the UI so probably you need to declare it only here

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,302 @@
+  provider: ProviderInfo,
+  containerConnectionInfo: ProviderContainerConnectionInfo,
+): void {
+  containerConnectionStatus.set(`${provider.name}-${containerConnectionInfo.name}`, {
```
probably need an helper that produces the key to be sure it's consistent over the class (one could use `-`, another `.`, etc)

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,302 @@
+    </p>
+  </div>
+  <div>
+    {#each $providerInfos as provider}
```
what is displayed if there are no provider ?

lstocchi replied:
In this PR i didn't add anything. In the next one it would be like this https://design.penpot.app/#/view/c1a43240-198d-11ed-817f-f940e3befd53?page-id=7524576d-080c-80dd-8002-0e62d894c2d4&section=interactions&index=5&share-id=611a6365-8047-804f-8002-0e636fc64875 .
Do you want me to add that part here?

benoitf replied:
I'm fine if it's coming just wanted to be sure that it'll come

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,302 @@
+$: containerConnectionDivHeight = 0;
+
+onMount(() => {
+  configurationProperties.subscribe(value => {
```
need to unsubscribe on the destroy lifecycle

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,331 @@
+import { onDestroy, onMount } from 'svelte';
+import type { IConfigurationPropertyRecordedSchema } from '../../../../main/src/plugin/configuration-registry';
+import { configurationProperties } from '../../stores/configurationProperties';
+import type { ContainerProviderConnection } from '@tmpwip/extension-api';
```
FYI we rename it to `@podman-desktop/api` last week so you'll probably need to rebase the code

benoitf reviewed (commented):
I tried the PR, good work 👍 
![image](https://user-images.githubusercontent.com/436777/224736520-84f9295e-1745-473c-b7ec-2c4140d910d4.png)

I noticed few things
I think for the size, we should not display unit twice
Create new button do nothing with Podman
I have external links icons (square with arrow) that does nothing
I'm missing 'starting'/'stopping' progress for when I start or stop a podman machine

For Docker:
-  I see the button create new while it's not possible
- I see vundefined (because the extension is not providing this optional value)


note sure the one that need to be fixed for this PR/follow up PR

deboer-tim commented:
> WDYM? Removing the GB next to the number?

I would remove it from the disk size and memory label. To me it is easier to read along with the number, and this won't cause problems if another machine was (e.g.) only 980Mb.

benoitf commented:
> I think for the size, we should not display unit twice

> WDYM? Removing the GB next to the number?
> immagine

I would do like in the mockup

![image](https://user-images.githubusercontent.com/436777/224742906-9bbada07-ca34-4169-ab92-b22719e672b3.png)

so we can make it human friendly (display 2TB instead of 200GB for example)

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,331 @@
+<script lang="ts">
```
we probably need a Component unit test for this new UI component

slemeur commented:
Tested it on my side (on mac):
- I think we are missing hover state and tooltips on the icons. 
- It seems some of the actions are not working on my side: "create new...", kebab menu, and "open"

I also have weird menu blinking.

See the recording:

https://user-images.githubusercontent.com/1636769/225367837-c3479810-fe62-4963-a2d6-56c205d72eff.mp4

lstocchi commented:
This is just a first part, then multiple PRs will come with everything you mentioned. I was not quite sure how to split the big refactoring so I started with this. 
Answering point by point 

> * I think we are missing hover state and tooltips on the icons.

I can work on this in this PR

> * It seems some of the actions are not working on my side: "create new...", kebab menu, and "open"

no, correct. Still not developed.

> * I think it would be nice to show a spinner when an action is processing (like stopping, starting or restarting)

It will be like what Mo said here https://github.com/containers/podman-desktop/issues/1337#issuecomment-1448653769

> I also have weird menu blinking.

the menu on the left? I guess it's an issue in the main branch. Worth opening a new issue

lstocchi commented:
I should have covered everything related to this PR. For actions that do not work, spinners, empty pages, i'll work on them in separate PRs.
Since last review i have added
1. hover style to buttons plus tooltip
2. remove old import
3. remove unit size label
4. hide `create new` button when no necessary
5. handle undefined version label
6. added a simple tests to verify buttons visibility

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.spec.ts:
```diff
@@ -0,0 +1,105 @@
+  providerInfo.containerConnections[0].status = 'started';
+  providerInfos.set([providerInfo]);
+  render(PreferencesResourcesRendering, {});
+  const buttons = screen.getAllByRole('button');
```
I guess we should have
```
const startButton = screen.getByRole('button', { name: 'Start' });
```

lstocchi replied:
I tried but it doesn't work that way. I don't know exactly what they mean by name but it's like it checks the text within the button. In my case i have an icon. 
I cannot use a title tag as it creates a white tooltip which has a design unrelated with the rest.
I used a name attribute for all buttons.

benoitf replied:
did you use the `aria-label` I think it's what should be used for the attribute name (so it works for a11y and testing-library)

lstocchi replied:
works fine with the `aria-label`. Thanks!!

benoitf commented:
FYI the 'Create new' button is not working on my side. (So I'm no longer able to create a Podman machine) so it's a regression from what we have now


https://user-images.githubusercontent.com/436777/225847138-ab067c8f-19d7-46d8-a724-67f73dbae983.mp4




![image](https://user-images.githubusercontent.com/436777/225847042-3c35becf-9a8d-4dfa-963d-372e5d55f2e5.png)

UI glitches:
the kebab menu is with a white color and hovering make it gray but as there is no actions maybe it should be disabled/removed

Disk Size is displayed with '100.00' GB I would expect '100 GB' 
For the memory it's displaying 2048.0 MB but I guess I should see 2GB

We are already including library like `filesize` to make it more 'human friendly'

The square/arrow top right icon should be removed or greyed out as it does nothing

lstocchi commented:
@benoitf your comments should be covered. 

- The `create new ...` button opens the old view. I'll refactor it in a new PR.
- I remove the three dots and disabled the arrow button (i kept it for a better visualization otherwise the buttons seem to be de-centered)
- I used the filesize lib for display sizes

benoitf commented:
I confirm I'm able to create podman machines

it seems long name is breaking layout
![image](https://user-images.githubusercontent.com/436777/225875079-6ea0eb99-845d-41aa-8eac-d71ba8898c02.png)


I have some horizontal lines being displayed sometimes
![image](https://user-images.githubusercontent.com/436777/225875987-1d7ed5f3-6c69-476a-8842-8fc77fe7eb69.png)

https://user-images.githubusercontent.com/436777/225875492-613cd55f-d9ad-4fd5-be6d-de52bca62b57.mp4

benoitf commented:
nit: (not a blocker to merge the PR) it seems the unit displayed is not the same unit that we set

for example if I do `podman machine init` 
```
  -m, --memory uint            Memory in MB (default 2048)
```

![image](https://user-images.githubusercontent.com/436777/225887807-334921d7-7696-4634-a666-5e1d72c19f1f.png)

but it looks it's not using the same base/unit as it displays 2.15GB and 107.37

```
> filesize.filesize(107374182400);
'107.37 GB'
> filesize.filesize(107374182400, {base: 2});
'100 GiB'
```

lstocchi commented:
> nit: (not a blocker to merge the PR) it seems the unit displayed is not the same unit that we set
> 
> for example if I do `podman machine init`
> 
> ```
>   -m, --memory uint            Memory in MB (default 2048)
> ```
> 
> ![image](https://user-images.githubusercontent.com/436777/225887807-334921d7-7696-4634-a666-5e1d72c19f1f.png)
> 
> but it looks it's not using the same base/unit as it displays 2.15GB and 107.37
> 
> ```
> > filesize.filesize(107374182400);
> '107.37 GB'
> > filesize.filesize(107374182400, {base: 2});
> '100 GiB'
> ```

In my case it reflects what i see if i list the machine in the terminal
![image](https://user-images.githubusercontent.com/49404737/225898100-77c63581-d6c5-46e4-a99b-525549777608.png)
![image](https://user-images.githubusercontent.com/49404737/225898149-68d9fc2a-cba9-44ee-84a4-e08e0fbdab7c.png)

but you say it makes sense as i create a machine with 100Gb of disk and 2Gb of memory as well.... why i don't see those values reflected in my machine list? 🤔

lstocchi commented:
Fixed separator borders and breaking div with long name

benoitf commented:
I guess podman machine ls is invalid :-)

```
$ podman machine list --format json
[
    {
...
        "Memory": "2147483648",
        "DiskSize": "107374182400",

    },
```

I'll file a bug in podman to see what should be displayed

lstocchi commented:
> I guess podman machine ls is invalid :-)
> 
> ```
> $ podman machine list --format json
> [
>     {
> ...
>         "Memory": "2147483648",
>         "DiskSize": "107374182400",
> 
>     },
> ```
> 
> I'll file a bug in podman to see what should be displayed

Maybe it's related to Windows as it's the same with format json

```
 {
        "Name": "podman-machine-default-super",
        "Default": false,
        "Created": "2023-03-17T11:49:20.6459695+01:00",
        "Running": false,
        "Starting": false,
        "LastUp": "2023-03-17T11:49:20.6459695+01:00",
        "Stream": "35",
        "VMType": "wsl",
        "CPUs": 0,
        "Memory": "0",
        "DiskSize": "535822336",
        "Port": 58740,
        "RemoteUsername": "user",
```

benoitf commented:
length is fixed 🎉  but dividers look still broken

https://user-images.githubusercontent.com/436777/225900302-87e095cf-3b9c-4cfd-88e7-aea07ab5f81c.mp4

benoitf commented on packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte:
```diff
@@ -0,0 +1,361 @@
+          bind:offsetWidth="{containerConnectionDivWidth}"
+          class="grow flex flex-wrap divide-gray-600">
+          {#each provider.containerConnections as container, containerIndex}
+            <div class="px-5 py-2 w-[240px] {calculateBorder(containerIndex, containerConnectionDivWidth)}">
```
AFAIK `calculateBorder` won't work. it's only called once

benoitf reviewed (approved):
so it works well except for dividers that are not being displayed correctly when we resize, have new machines being added/removed

![image](https://user-images.githubusercontent.com/436777/225902742-41ee66e8-5c6f-422d-b11e-89a8d883f97b.png)

lstocchi commented:
Just to clarify.
Is this broken for you or is it ok?
![image](https://user-images.githubusercontent.com/49404737/225903418-dcea194a-3e1d-4cfa-a0a1-2a524e5c0570.png)
Would you expect the bottom border for the second/third element in the first row?

benoitf commented:
I feel it's broken but maybe we can have thoughts from @mairin

lstocchi commented:
> @lstocchi how does it look if you just kill all the borders besides the far left vert one?

Something like this?
![demo-borders](https://user-images.githubusercontent.com/49404737/225908112-add19754-aa8b-4961-943d-1e14d62540de.gif)

mairin commented:
That looks great to me, and keeps things simple. what do you think?

lstocchi commented:
> That looks great to me, and keeps things simple. what do you think?

With many machines and a small screen (so several machines on multiple rows) i prefer the separator as it gives a clearer view but i wonder how many users will work with such a number. 
We can start without borders and then add them in future if we think it can improve the ui.

benoitf commented:
I like boxes/separators (maybe could be slightly different background color for the widget) but I'm fine to keep it like this for the first iteration as it may feel that we miss something but it does not feel 'broken'

so I'm +1 to keep it simple

lstocchi commented:
Merging it so we can proceed with next PRs. All remaining things will be fixed soon. Refactoring on progress :)
